Medium perturbation

dbochkov-flexcompute opened the pull request:
Heat and charge perturbation functionality. Currently, implemented for `Medium` and `PoleResidue`, but this is ready for an initial "big picture" review, to check if anything seems bad/needs a rework. Here's a notebook with demonstrations and explanations to help the review https://github.com/flexcompute-readthedocs/tidy3d-docs/blob/daniil/perturbation-medium/docs/source/notebooks/PerturbationMedium.ipynb. If everything looks more or less alright, I will go ahead and finish some remaining to-do's:
- [ ] do the rest of dispersive models, that is, add `PerturbationDebye`, `PerturbationLorentz`, etc
- [x] add front-end tests
- [x] check all code documentation is in place

tagging @momchil-flex just in case

tylerflex commented on tidy3d/components/parameter_perturbation.py:
```diff
@@ -0,0 +1,722 @@
+            # if temperature range not present just return original function
+            if np.any(hole_density < h_min) or np.any(hole_density > h_max):
+                log.warning(
+                    "Hole density values passed to 'ChargePerturbation.sample()'"
```
noticed some inconsistency regarding periods at end of messages and docstring descriptions. let's try to add them to all.

tylerflex commented on tidy3d/components/parameter_perturbation.py:
```diff
@@ -0,0 +1,722 @@
+
+    temperature_ref: pd.NonNegativeFloat = pd.Field(
+        ...,
+        title="Reference temperature",
```
I think some of the docstrings and descriptions could use a bit more expanding on. To someone not familiar with this, it could be hard to understand what these classes and their fields do. For example, the docstring "Linear heat perturbation" could be "Specification describing how heat quantities affect optical medium properties." and the description of "reference temperature" could be something like "temperature at which the thermo optic coefficient is valid. Away from this temperature ..."

Your notebook has a lot of really useful descriptions, maybe some can be copied into the docs too.

tylerflex commented on tidy3d/components/parameter_perturbation.py:
```diff
@@ -0,0 +1,722 @@
+        ...,
+        title="Thermo-optic Coefficient",
+        description="Thermo-optic Coefficient.",
+    )
```
units? seems a lot of the other ones might be missing them as well

dbochkov-flexcompute replied:
the units are actually unknown here because it can be applied to any parameter, but I set it to `1/K` as in case of a dimensionless parameter

tylerflex commented on tidy3d/components/parameter_perturbation.py:
```diff
@@ -0,0 +1,722 @@
+
+
+ChargePerturbationType = Union[LinearChargePerturbation, CustomChargePerturbation]
+
```
remove extra newlines?

tylerflex commented on tidy3d/components/simulation.py:
```diff
@@ -2787,3 +2790,142 @@ def allow_gain(self) -> bool:
+
+    @staticmethod
+    def _sel_data_inside_bounds(data: SpatialDataArray, bounds: Bound) -> Tuple[bool, SpatialDataArray]:
+
```
docstring

tylerflex commented on tidy3d/components/simulation.py:
```diff
@@ -2787,3 +2790,142 @@ def allow_gain(self) -> bool:
         return False
+
+    @staticmethod
+    def _sel_data_inside_bounds(data: SpatialDataArray, bounds: Bound) -> Tuple[bool, SpatialDataArray]:
```
a few things that might help make this more 

1. Maybe this should be a DataArray method? to avoid Simulation getting even larger?
2. does it make sense to separate `is_full_cover` and `sel_data_inside_bounds` into two functions? for example `sel_data_inside_bounds` does the data selection and `is_full_cover` just looks at bounds compared to self's coordinates?

dbochkov-flexcompute replied:
imlpemented

tylerflex commented on tidy3d/components/simulation.py:
```diff
@@ -2787,3 +2790,142 @@ def allow_gain(self) -> bool:
+        electron_density: SpatialDataArray = None,
+        hole_density: SpatialDataArray = None,
+    ) -> Simulation:
+        """Apply heat and/or charge data to perturbation mediums contained in the Simulation
```
let's make sure the users know that this returns a copy, so they dont do
```py
sim.apply_perturbations(...)
```
and not realize why things didn't change.  maybe instead of "Apply heat and/or charge data to ..." we should word it as "Return a copy of the simulation with heat and/or charge data applied to ...

dbochkov-flexcompute replied:
Agree. I also renamed it to `.perturbed_mediums_copy()` by analogy to `.updated_copy()`

tylerflex commented on tidy3d/components/simulation.py:
```diff
@@ -2787,3 +2790,142 @@ def allow_gain(self) -> bool:
+                if not np.all([t_cover, e_cover, h_cover]):
+                    log.warning(
+                        "Provided temperature, electron density, and/or hole density data does not "
+                        f"fully cover structure {s_ind}."
```
maybe
```
"fully cover 'structures[{s_ind}]'."
```

tylerflex commented on tidy3d/components/simulation.py:
```diff
@@ -2787,3 +2790,142 @@ def allow_gain(self) -> bool:
+        new_structures = []
+        for s_ind, s in enumerate(structures):
+            med = s.medium
+            if isinstance(med, AbstractPerturbationMedium):
```
could this be factored into a `Structure.apply_perturbations` method? to simplify?

And for the `Simulation.medium` part, call `Medium.apply_perturbations`?

dbochkov-flexcompute replied:
it could be, but we would need to pass simulation bounds `sim_bounds` to it and method would have to return two values: new structure + flag whether coverage was full. So, maybe it fine to keep it all exposed

tylerflex replied:
ok

tylerflex commented on tidy3d/plugins/dispersion/fit.py:
```diff
@@ -333,6 +333,7 @@ def _make_medium(self, coeffs):
     def _fit_single(
         self,
         num_poles: int = 3,
+        guess = None,
```
do we want to add this to the user-facing `fit` method?

dbochkov-flexcompute replied:
added

tylerflex reviewed (changes requested):
Thanks Daniil, this is a ton of work and I think it is really well organized and covers everything I can think of very well. Most of my comments are just refining it for public consumption. The notebook is also extremely useful and will be invaluable. There were a few spelling issues, I recommend this https://pypi.org/project/jupyterlab-spellchecker/ to catch some of them before we do a more thorough proofread.

Thanks!

weiliangjin2021 commented on tidy3d/components/parameter_perturbation.py:
```diff
@@ -0,0 +1,722 @@
+
+        t_range = self.temperature_range
+        temperature_clip = np.clip(temperature, t_range[0], t_range[1])
+        return self.perturbation_values.interp(T=temperature_clip).drop_vars("T")
```
shall we have "linear" and "nearest" option here?

dbochkov-flexcompute replied:
yeah, added for completeness

weiliangjin2021 reviewed (commented):
Overall looks quite good! A few small comments.

dbochkov-flexcompute commented on tidy3d/components/medium.py:
```diff
@@ -2927,6 +2928,303 @@ class CustomAnisotropicMediumInternal(CustomAnisotropicMedium):
+            Medium specification after application of heat and/or charge data.
+        """
+
+        if all([x is None for x in [temperature, electron_density, hole_density]]):
```
it seems to have issues when elements are numpy arrays 
```
>>> any([np.array([1, 2]), np.array([3, 4])])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
```
thus, decided to use explicit `is not None`

tylerflex replied:
oh ok, sounds good

dbochkov-flexcompute commented on tidy3d/components/medium.py:
```diff
@@ -2927,6 +2928,303 @@ class CustomAnisotropicMediumInternal(CustomAnisotropicMedium):
+    def warn_permittivity_range(cls, val, values):
+        """Warn if permittivity can become less than one due to heat/charge perturbations."""
+
+        if val:
```
good idea, implemented

dbochkov-flexcompute commented on tidy3d/components/data/data_array.py:
```diff
@@ -447,6 +447,33 @@ class TriangleMeshDataArray(DataArray):
+    Example
+    -------
+    >>> T = [0, 1e-12, 2e-12]
+    >>> td = HeatDataArray((1+1j) * np.random.random((3,)), coords=dict(T=T))
```
temperature here is a coordinate f=f(T), so values could be real or complex

dbochkov-flexcompute commented on tidy3d/components/parameter_perturbation.py:
```diff
@@ -0,0 +1,722 @@
+
+    @cached_property
+    @abstractmethod
+    def perturbation_range(self) -> Union[Tuple[float, float], Tuple[Complex, Complex]]:
```
perturbations to poles can be complex

dbochkov-flexcompute commented on tidy3d/components/parameter_perturbation.py:
```diff
@@ -0,0 +1,722 @@
+        description="2D array of perturbation values.",
+    )
+
+    temperature_range: Tuple[pd.NonNegativeFloat, pd.NonNegativeFloat] = pd.Field(
```
correct, just a copy-paste leftover

dbochkov-flexcompute commented on tidy3d/components/parameter_perturbation.py:
```diff
@@ -0,0 +1,722 @@
+            log.warning(
+                "Temperature range for 'CustomHeatPerturbation' is calculated automatically "
+                "based on provided 'perturbation_values'. Provided 'temperature_range' will be overwritten."
+            )
```
specified range here is not used at all, it always if overwritten by range from data

weiliangjin2021 replied:
in that case, it should be property rather than a field?

weiliangjin2021 replied:
i see, then it looks fine.

dbochkov-flexcompute commented:
big cleanup, added many docstrings, added tests, and fixed many loose ends. I am considering adding rest of dispersive models `PerturbationDebye`, `PerturbationLorentz`, etc, later, and merging this without them. What do you think?

weiliangjin2021 commented:
> big cleanup, added many docstrings, added tests, and fixed many loose ends. I am considering adding rest of dispersive models `PerturbationDebye`, `PerturbationLorentz`, etc, later, and merging this without them. What do you think?

Agree, as long as we can have a good heat example in the current stage.

dbochkov-flexcompute commented on tidy3d/components/validators.py:
```diff
@@ -249,3 +251,88 @@ def _single_frequency_in_range(cls, val: FieldDataset, values: dict) -> FieldDat
+                            raise SetupError(
+                                f"Provided '{field_name}' contains parameter perturbation  is "
+                                f"provided for a 'None' base value of '{base_field_name}'."
+                            )
```
Yeah, this is redundant, removed

dbochkov-flexcompute commented on tidy3d/components/medium.py:
```diff
@@ -2927,6 +2928,256 @@ class CustomAnisotropicMediumInternal(CustomAnisotropicMedium):
+        allowed_imag_range=[None],
+        allowed_complex=False,
+    )
+    _conductivity_perturbation_validator = validate_parameter_perturbation(
```
so, these validators only warn if conductivity/permittivity can potentially go out of bounds (within the provided/calculated `temperature_range`, `electron_range`, and `hole_range`). The strict validation is left to the corresponding `CustomMedium`. So perhaps it's still ok to warn the user about potentially negative conductivity even if `allow_gain=True`?

weiliangjin2021 replied:
I see, so it will warn and then error. In that regard, there seems to be missing `allow_gain` in the return of 
```
    return CustomMedium(
            permittivity=permittivity_field,
            conductivity=conductivity_field,
            name=self.name,
            subpixel=True,
        )
```
and it seems better to add `subpixel` as a field of PerturbedMedium so that one can choose to apply subpixel or not. But warn that for a uniform medium, supixel parameter will be ignored.

dbochkov-flexcompute replied:
Oh, you're right, thanks for catching that. Actually, `frequency_range` should also be passed. Fixed it.

dbochkov-flexcompute commented:
if there are no other comments at this point, is it ok to merge this into pre/2.4?

tylerflex commented:
Yea I've given all of my comments. I think it's ok if everyone else does

weiliangjin2021 commented:
Green light from me.
